### DISCUSSION TITLE
Python errors in benchmark toolset

### DISCUSSION THREAD

**Discussion by bdw429s** [TechEmpower/FrameworkBenchmarks#6338] [General]
Hi all, I'm running the benchmarks for a single framework on my Windows machine.  I have the latest `master` branch pulled and I'm running the tests via Docker Machine like so

```
$ docker run -it --rm --network tfb -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks techempower/tfb --mode verify --test cfml
```
The test verifications themselves run fine, but this Python error halts execution at the end 

```
...
No problems to report
wrk: Build time: 1s
techempower/postgres:latest: Build time: 1s
cfml: Build time: 1s
cfml: Time starting database: 1s
cfml: Time until accepting requests: 3s
cfml: Verify time: 57s
A fatal error has occurred
Traceback (most recent call last):
  File "/FrameworkBenchmarks/toolset/run-tests.py", line 244, in main
    any_failed = benchmarker.run()
  File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 67, in run
    if not self.__run_test(test, benchmark_log):
  File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 247, in __run_test
    self.docker_helper.stop()
  File "/FrameworkBenchmarks/toolset/utils/docker_helper.py", line 284, in stop
    self.database.containers.prune()
  File "/usr/local/lib/python2.7/dist-packages/docker/models/containers.py", line 951, in prune
    return self.client.api.prune_containers(filters=filters)
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 34, in wrapper
    return f(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/container.py", line 971, in prune_containers
    return self._result(self._post(url, params=params), True)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 267, in _result
    self._raise_for_status(response)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 263, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
APIError: 409 Client Error: Conflict ("a prune operation is already running")
Shutting down (may take a moment)
```
Furthermore, if I run the tests themselves (without the `--mode verify` flags) like so...
```
$ docker run -it --rm --network tfb -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks techempower/tfb --test cfml
```
The tests run ok but I get the same error as above plus this one below:
```
...
STARTTIME 1612477645
ENDTIME 1612477661
cfml: Traceback (most recent call last):
cfml:   File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 209, in __run_test
cfml:     self.__benchmark(test, benchmark_log)
cfml:   File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 295, in __benchmark
cfml:     benchmark_type(test_type)
cfml:   File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 285, in benchmark_type
cfml:     results = self.results.parse_test(framework_test, test_type)
cfml:   File "/FrameworkBenchmarks/toolset/utils/results.py", line 149, in parse_test
cfml:     rawData["startTime"], rawData["endTime"], 1)
cfml:   File "/FrameworkBenchmarks/toolset/utils/results.py", line 471, in __parse_stats
cfml:     if len(main_header[item_num]) != 0:
cfml: IndexError: list index out of range
cfml: Error during test: cfml
```

In the examples above, I'm testing the new `cfml` framework, but I get the same Python errors even if I run the commands against an existing framework such as `gemini` so it doesn't appear to be related to the framework being tested.

**Comment by bdw429s** [comment]
Any ideas on this?

**Comment by msmith-techempower** [comment]
This is a strange one... it looks like the Docker daemon is already attempting to perform a prune operation while the toolset is requesting it.

Can you try restarting the Docker daemon and testing again?

**Reply by bdw429s** [reply]
Thanks for the reply @msmith-techempower  Restarting my Docker machine does seem to have cleared up the "prune" error.  However, the `list index out of range` error still persists.

**Comment by bdw429s** [comment]
The `list index out of range` Python error is still a blocker preventing me from effectively run the benchmarks locally.

**Reply by msmith-techempower** [reply]
I honestly don't have any good ideas here - I went through the stack trace and it appears like the `dstat` data is not captured correctly. This error is being thrown when trying to parse the headers from `dstat`... 

can you check that `/results/[timestamp]/test_type/test_name/stats.txt` exists, and if so post it here?

**Reply by bdw429s** [reply]
I'm not sure which test name the code was parsing, but the last one to run before the error was `db`.  I'll attach that stats file and I can provide the others if they help. ( `FrameworkBenchmarks\results\20210217181514\cfml\db\stats.txt` )

[stats.txt](https://github.com/TechEmpower/FrameworkBenchmarks/files/5998019/stats.txt)

**Reply by msmith-techempower** [reply]
I don't see anything broken about this... the headers are clearly there, so I'm not sure why the toolset is erroring on parsing them...

**Reply by bdw429s** [reply]
I'm not a Python dev (and I have no desire to become one) but I can at least try to put in some debugging code to see what's going on.  However, is there a way to run just this routine directly?  The only way I can create this error is by running the entire toolset for a language, and waiting 5 minutes after each line of code to see if it works doesn't sound very fun :)

**Reply by bdw429s** [reply]
@msmith-techempower Actually, I see something that may be the issue.  The code appears to be looping over all the items in the rows and then accessing the corresponding header name from the headers.  This assumes that the header list is as long (or longer) than the number of items in the row.  You mentioned above that the headers were present, but even if they were present, if they contained fewer items than the row, that would also cause this error.  I opened the stats file (minus the first 4 rows) in open office and look what I see:

![image](https://user-images.githubusercontent.com/584009/108387377-2653db00-71d3-11eb-855b-c9609ea8570b.png)

The header row is two columns shorter than the data rows.  I confirmed that `zram0` is the last header.  I think your issue is that there are fewer headers than items in the rows and the code assumes there will be a header for each row.

**Reply by bdw429s** [reply]
@msmith-techempower Does the findings I posted here last week regarding the CSV file having mismatched header and data rows help you understand the error?

**Reply by bdw429s** [reply]
@msmith-techempower Would you like me to create a ticket for my finding above where the column headers in the CSV don't match value rows?

**Reply by bdw429s** [reply]
@msmith-techempower Would you like a ticket for this finding?

**Reply by bdw429s** [reply]
It's been a month since I've gotten a reply here, so I created a ticket here

https://github.com/TechEmpower/FrameworkBenchmarks/issues/6464
